### PR TITLE
implement timedelta ISO8601 duration format encoder

### PR DESCRIPTION
**Pull request by cfkanesan**
Add JSON serialization support for `datetime.timedelta` objects

Based on https://stackoverflow.com/questions/27168175/convert-a-datetime-timedelta-into-iso-8601-duration-in-python

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Sorry, I really dislike ISO8601 duration format, also it's not trivial to parse in javascript.

If we can make it optional I'd be ok with adding the option but I don't want to make it the default.

**Comment by samuelcolvin** [comment]
so I guess:
* add the method
* add  `encoder=None` argument to `json()` which defaults to `pydantic_encoder`
* then document using a custom encoder and using `iso8601_duration`

**Comment by cfkanesan** [comment]
Sounds good to me.

Awesome library by the way :+1:

**Comment by samuelcolvin** [comment]
Thank you. 

Would be great if you could work on this?

File: pydantic/json.py [hunk 1]
```diff
@@ -13,11 +13,22 @@ def isoformat(o):
...
+    seconds = round(seconds, 6)
+    return f'P{days}DT{hours}H{minutes}M{seconds}S'
+
+
 ENCODERS_BY_TYPE = {
     UUID: str,
     datetime.datetime: isoformat,
     datetime.date: isoformat,
     datetime.time: isoformat,
+    datetime.timedelta: iso8601_duration,
```
Comment (samuelcolvin) [inline pydantic/json.py]: I guess the default should just be `lambda td: f'{td.total_seconds():0.3f}'`

**Comment by samuelcolvin** [comment]
@cfkanesan when do you expect to work on this?

Sorry to hassle, just would be good to get this fixed. Happy to do it myself or leave it to you as you wish.

File: pydantic/json.py (hunk 1 above)
Comment (cfkanesan) [reply pydantic/json.py]: I think it's a good idea except that timedelta has a resolution of up to 1 microsecond, I would suggest formatting to `:0.6f`

**Comment by cfkanesan** [comment]
Sorry, I don't have much time to myself these days. I'll try to give it a go.

Ideally, I would like to have the option to define (de)serialization methods per type (as in by subclassing the type in question) but maybe the more sensible route is to call the `dict()` method and pass the result to a custom JSON Encoder.

**Comment by samuelcolvin** [comment]
Thank you this contributed a lot to #247.

**Pull request closed without merging**